daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/ApiCommand.scala

```diff
@@ -84,3 +84,29 @@ case class ApiCommands(
     ledgerEffectiveTime: Time.Timestamp,
     commandsReference: String,
 )
+
+/** An additional contract that is used to resolve contract id and contract key lookups during interpretation.
+  *
+  * @param   templateId   identifier of the template of disclosed contract
+  * @param   contractId   the contract id of the disclosed contract
+  * @param   argument     the payload of the disclosed contract
+  * @param   metadata     metatdata attached to this disclosure. See [ContractMetadata].
+  */
+final case class DisclosedContract(
+    templateId: Identifier,
+    contractId: Value.ContractId,
+    argument: Value.VersionedContractInstance,
```

Comment:
I think this should just be a daml-lf value not a versioned contract instance. The participant does not know the version. wdyt @rautenrieth-da?

Reply:
The return value could be a contract instance though. Maybe we want this to be DisclosedContract and then we return Versioned[DisclosedContract]?

Reply:
The less data the engine takes, the better. The GRPC submission call only has an unversioned value and the template id, and we don't want to look up the version outside of the engine.

We could use `Value.VersionedContractInstance` throughout the codebase, and carefully document where the agreement text and/or the transaction version don't need to contain valid data. I think I would prefer however if we changed all interfaces to only take the data that is really needed. Using `DisclosedContract` which contains an unversioned value and then returning a `Versioned[DisclosedContract]` sounds like a good idea to me.

Small note, `Value.VersionedContractInstance` already contains the template id. If we stick to `Value.VersionedContractInstance`, we don't need the extra `templateId` field.

Reply:
This should not be a versioned contract instance, as it is not normalized/typechecked.

Reply:
returning a `Versioned[DisclosedContract]` seems a good idea to me.

Reply:
> We could use Value.VersionedContractInstance throughout the codebase, and carefully document where the agreement text and/or the transaction version don't need to contain valid data.

I would rather implement two different type that implement a same sealed abstract class, or changing the type of agreement text to `Option[String]`.

Reply:
I am fine with change the type of `argument` in another PR.